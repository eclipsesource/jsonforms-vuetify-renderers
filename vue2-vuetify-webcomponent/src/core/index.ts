export * from './dynamicDefaults';
export * from './i18n';
export * from './json-refs';
export * from './keywords';
export * from './transform';
export * from './types';
export * from './validate';
