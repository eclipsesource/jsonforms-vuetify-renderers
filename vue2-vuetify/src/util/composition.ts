import {
  composePaths,
  computeLabel,
  getFirstPrimitiveProp,
  isDescriptionHidden,
  JsonFormsSubStates,
  Resolve,
} from '@jsonforms/core';
import cloneDeep from 'lodash/cloneDeep';
import merge from 'lodash/merge';
import { useStyles } from '../styles';
import { computed, ComputedRef, inject, ref } from '../vue';
import { Ajv } from 'ajv';

const useControlAppliedOptions = <I extends { control: any }>(input: I) => {
  return computed(() =>
    merge(
      {},
      cloneDeep(input.control.value.config),
      cloneDeep(input.control.value.uischema.options)
    )
  );
};

const useComputedLabel = <I extends { control: any }>(
  input: I,
  appliedOptions: ComputedRef<any>
) => {
  return computed((): string => {
    return computeLabel(
      input.control.value.label,
      input.control.value.required,
      !!appliedOptions.value?.hideRequiredAsterisk
    );
  });
};

/**
 * Adds styles, isFocused, appliedOptions and onChange
 */
export const useVuetifyControl = <
  I extends { control: any; handleChange: any }
>(
  input: I,
  adaptValue: (target: any) => any = (v) => v
) => {
  const appliedOptions = useControlAppliedOptions(input);

  const isFocused = ref(false);
  const onChange = (value: any) => {
    input.handleChange(input.control.value.path, adaptValue(value));
  };

  const persistentHint = (): boolean => {
    return !isDescriptionHidden(
      input.control.value.visible,
      input.control.value.description,
      isFocused.value,
      !!appliedOptions.value?.showUnfocusedDescription
    );
  };

  const computedLabel = useComputedLabel(input, appliedOptions);

  const controlWrapper = computed(() => {
    const { id, description, errors, label, visible, required } =
      input.control.value;
    return { id, description, errors, label, visible, required };
  });

  const styles = useStyles(input.control.value.uischema);

  return {
    ...input,
    styles,
    isFocused,
    appliedOptions,
    controlWrapper,
    onChange,
    persistentHint,
    computedLabel,
  };
};

/**
 * Adds styles and appliedOptions
 */
export const useVuetifyLayout = <I extends { layout: any }>(input: I) => {
  const appliedOptions = computed(() =>
    merge(
      {},
      cloneDeep(input.layout.value.config),
      cloneDeep(input.layout.value.uischema.options)
    )
  );
  return {
    ...input,
    styles: useStyles(input.layout.value.uischema),
    appliedOptions,
  };
};

/**
 * Adds styles, appliedOptions and childUiSchema
 */
export const useVuetifyArrayControl = <I extends { control: any }>(
  input: I
) => {
  const appliedOptions = useControlAppliedOptions(input);

  const computedLabel = useComputedLabel(input, appliedOptions);

  const childLabelForIndex = (index: number) => {
    const childLabelProp =
      input.control.value.uischema.options?.childLabelProp ??
      getFirstPrimitiveProp(input.control.value.schema);
    if (!childLabelProp) {
      return `${index}`;
    }
    const labelValue = Resolve.data(
      input.control.value.data,
      composePaths(`${index}`, childLabelProp)
    );
    if (
      labelValue === undefined ||
      labelValue === null ||
      Number.isNaN(labelValue)
    ) {
      return '';
    }
    return `${labelValue}`;
  };
  return {
    ...input,
    styles: useStyles(input.control.value.uischema),
    appliedOptions,
    childLabelForIndex,
    computedLabel,
  };
};

/**
 * Adds styles and appliedOptions
 */
export const useVuetifyBasicControl = <I extends { control: any }>(
  input: I
) => {
  const appliedOptions = useControlAppliedOptions(input);

  return {
    ...input,
    styles: useStyles(input.control.value.uischema),
    appliedOptions,
  };
};

/**
 * Extracts Ajv from JSON Forms
 */
export const useAjv = () => {
  const jsonforms = inject<JsonFormsSubStates>('jsonforms');

  if (!jsonforms) {
    throw new Error(
      "'jsonforms' couldn't be injected. Are you within JSON Forms?"
    );
  }

  // should always exist
  return jsonforms.core?.ajv as Ajv;
};
