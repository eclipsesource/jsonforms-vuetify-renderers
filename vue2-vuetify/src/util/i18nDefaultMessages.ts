export const i18nDefaultMessages = {
  arraylayout: {
    add: 'Add',
    delete: 'Delete',
    moveUp: 'Move Up',
    moveDown: 'Move Down',
    dialogTitle: 'Delete {{ element }}?',
    dialogText: 'The element will be deleted.',
    dialogConfirm: 'Delete',
    dialogCancel: 'Cancel',
  },
};
