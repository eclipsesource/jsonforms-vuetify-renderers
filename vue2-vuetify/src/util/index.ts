export * from './options';
export * from './composition';
export * from './validator';
export * from './datejs';
export * from './i18nDefaultMessages';
