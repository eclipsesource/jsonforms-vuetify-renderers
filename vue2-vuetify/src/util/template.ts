import templateFn from 'lodash/template';
import merge from 'lodash/merge';
import { TemplateExecutor, TemplateOptions } from 'lodash';

const interpolate = /{{([\s\S]+?)}}/g;

export const template = (
  templateString?: string,
  options?: TemplateOptions
): TemplateExecutor => {
  return templateFn(
    templateString,
    merge({ interpolate: interpolate }, options)
  );
};
