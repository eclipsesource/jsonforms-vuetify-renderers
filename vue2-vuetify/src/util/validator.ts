import { Options } from 'ajv';
import { createAjv as createAjvCore } from '@jsonforms/core';
import Ajv from 'ajv';

export const createAjv = (options?: Options): Ajv => {
  const ajv = createAjvCore(options);
  ajv.addFormat('password', (_) => true);
  return ajv;
};
