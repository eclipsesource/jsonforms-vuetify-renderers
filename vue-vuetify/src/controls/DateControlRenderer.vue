<template>
  <control-wrapper
    v-bind="controlWrapper"
    :styles="styles"
    :isFocused="isFocused"
    :appliedOptions="appliedOptions"
  >
    <v-text-field
      :id="control.id + '-input'"
      :class="styles.control.input"
      :disabled="!control.enabled"
      :autofocus="appliedOptions.focus"
      :placeholder="appliedOptions.placeholder"
      :label="computedLabel"
      :hint="control.description"
      :persistent-hint="persistentHint()"
      :required="control.required"
      :error-messages="control.errors"
      v-bind="vuetifyProps('v-text-field')"
      :model-value="control.data"
      @update:model-value="onChange"
      @focus="isFocused = true"
      @blur="handleBlur"
      type="date"
    />
  </control-wrapper>
</template>

<script lang="ts">
import {
  ControlElement,
  isDateControl,
  JsonFormsRendererRegistryEntry,
  rankWith,
} from '@jsonforms/core';
import { defineComponent } from 'vue';

import {
  rendererProps,
  RendererProps,
  useJsonFormsControl,
} from '@jsonforms/vue';
import { VTextField } from 'vuetify/components';
import { useVuetifyControl } from '../util';
import { default as ControlWrapper } from './ControlWrapper.vue';

const controlRenderer = defineComponent({
  name: 'date-control-renderer',
  components: {
    ControlWrapper,
    VTextField,
  },
  props: {
    ...rendererProps<ControlElement>(),
  },
  setup(props: RendererProps<ControlElement>) {
    const adaptValue = (value: any) => value || undefined;
    const control = useVuetifyControl(useJsonFormsControl(props), adaptValue);
    return { ...control, adaptValue };
  },
});

export default controlRenderer;

export const entry: JsonFormsRendererRegistryEntry = {
  renderer: controlRenderer,
  tester: rankWith(2, isDateControl),
};
</script>
