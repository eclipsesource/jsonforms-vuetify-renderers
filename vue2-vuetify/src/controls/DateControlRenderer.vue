<template>
  <control-wrapper
    v-bind="controlWrapper"
    :styles="styles"
    :isFocused="isFocused"
    :appliedOptions="appliedOptions"
  >
    <v-text-field
      type="date"
      :id="control.id + '-input'"
      :class="styles.control.input"
      :disabled="!control.enabled"
      :autofocus="appliedOptions.focus"
      :placeholder="appliedOptions.placeholder"
      :label="computedLabel"
      :hint="control.description"
      :persistent-hint="persistentHint()"
      :required="control.required"
      :error-messages="control.errors"
      :value="control.data"
      @change="onChange"
      @focus="isFocused = true"
      @blur="isFocused = false"
    />
  </control-wrapper>
</template>

<script lang="ts">
import {
  ControlElement,
  JsonFormsRendererRegistryEntry,
  rankWith,
  isDateControl,
} from '@jsonforms/core';
import { defineComponent } from '../vue';
import {
  rendererProps,
  useJsonFormsControl,
  RendererProps,
} from '@jsonforms/vue2';
import { default as ControlWrapper } from './ControlWrapper.vue';
import { useVuetifyControl } from '../util';
import { VTextField } from 'vuetify/lib';

const controlRenderer = defineComponent({
  name: 'date-control-renderer',
  components: {
    ControlWrapper,
    VTextField,
  },
  props: {
    ...rendererProps<ControlElement>(),
  },
  setup(props: RendererProps<ControlElement>) {
    return useVuetifyControl(useJsonFormsControl(props));
  },
});

export default controlRenderer;

export const entry: JsonFormsRendererRegistryEntry = {
  renderer: controlRenderer,
  tester: rankWith(2, isDateControl),
};
</script>
