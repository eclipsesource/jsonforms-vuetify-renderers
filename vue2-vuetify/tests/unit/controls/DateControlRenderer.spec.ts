import { clearAllIds } from '@jsonforms/core';
import { Wrapper } from '@vue/test-utils';
import DateControlRenderer, {
  entry as dateControlRendererEntry,
} from '../../../src/controls/DateControlRenderer.vue';
import { wait } from '../../../tests';
import { mountJsonForms } from '../util';

describe('DateControlRenderer.vue', () => {
  const renderers = [dateControlRendererEntry];

  const data = '2021-03-09';
  const schema = {
    type: 'string',
    title: 'My Date',
    format: 'date',
  };
  const uischema = {
    type: 'Control',
    scope: '#',
    options: {
      placeholder: 'date placeholder',
    },
  };

  let wrapper: Wrapper<any, Element>;

  beforeEach(() => {
    // clear all ids to guarantee that the snapshots will always be generated with the same ids
    clearAllIds();
    wrapper = mountJsonForms(data, schema, renderers, uischema);
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('check if child DateControlRenderer exists', () => {
    expect(wrapper.getComponent(DateControlRenderer));
  });

  it('renders a date input', () => {
    expect(wrapper.find('input[type="date"]').exists()).toBe(true);
  });

  it('renders title as label', () => {
    expect(wrapper.find('label').text()).toEqual('My Date');
  });

  it('emits a data change', async () => {
    const input = wrapper.find('input[type="date"]');
    await input.setValue('2021-03-10');
    await input.trigger('blur');
    // 300 ms debounceWait
    await wait(300);
    expect(wrapper.vm.$data.data).toEqual('2021-03-10');
  });

  it('should have a placeholder', async () => {
    const input = wrapper.find('input[type="date"]');
    // select the input so the placeholder is generated
    await input.trigger('click');
    const placeholder = input.attributes('placeholder');
    expect(placeholder).toEqual('date placeholder');
  });

  it('should render component and match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
