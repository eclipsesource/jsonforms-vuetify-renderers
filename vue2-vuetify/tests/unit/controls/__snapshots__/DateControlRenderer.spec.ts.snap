// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateControlRenderer.vue should render component and match snapshot 1`] = `
<div data-app="true"
     class="v-application v-application--is-ltr theme--light"
     id="app"
>
  <div class="v-application--wrap">
    <div id="#"
         class="control"
         appliedoptions="[object Object]"
         errors
         label="My Date"
    >
      <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--placeholder input">
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-text-field__slot">
              <label for="#-input"
                     class="v-label v-label--active theme--light"
                     style="left: 0px; position: absolute;"
              >
                My Date
              </label>
              <input id="#-input"
                     type="text"
              >
            </div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <transition-group-stub tag="div"
                                     name="message-transition"
                                     class="v-messages__wrapper"
              >
              </transition-group-stub>
            </div>
          </div>
        </div>
      </div>
      <div class="v-menu">
      </div>
    </div>
  </div>
</div>
`;

exports[`DateControlRenderer.vue should render component and match snapshot when clicked 1`] = `
<div data-app="true"
     class="v-application v-application--is-ltr theme--light"
     id="app"
>
  <div class="v-application--wrap">
    <div id="#"
         class="control"
         appliedoptions="[object Object]"
         errors
         label="My Date"
    >
      <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--placeholder input">
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-text-field__slot">
              <label for="#-input"
                     class="v-label v-label--active theme--light"
                     style="left: 0px; position: absolute;"
              >
                My Date
              </label>
              <input id="#-input"
                     type="text"
              >
            </div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <transition-group-stub tag="div"
                                     name="message-transition"
                                     class="v-messages__wrapper"
              >
              </transition-group-stub>
            </div>
          </div>
        </div>
      </div>
      <div class="v-menu">
        <transition-stub name="scale-transition">
        </transition-stub>
      </div>
    </div>
  </div>
  <div role="menu"
       class="v-menu__content theme--light menuable__content__active "
       style="max-height: auto; min-width: 290px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 8; display: none;"
  >
    <div class="v-picker v-card v-picker--date theme--light">
      <div class="v-picker__title primary">
        <div class="v-date-picker-title">
          <div class="v-picker__title__btn v-date-picker-title__year">
            2021
          </div>
          <div class="v-picker__title__btn v-date-picker-title__date v-picker__title__btn--active">
            <transition-stub name="picker-transition">
              <div>
                Tue, Mar 9
              </div>
            </transition-stub>
          </div>
        </div>
      </div>
      <div class="v-picker__body theme--light"
           style="width: 290px;"
      >
        <transition-stub name="fade-transition">
          <div>
            <div class="v-date-picker-header theme--light">
              <button type="button"
                      class="v-btn v-btn--icon v-btn--round theme--light v-size--default"
                      aria-label="Previous month"
              >
                <span class="v-btn__content">
                  <i aria-hidden="true"
                     class="v-icon notranslate mdi mdi-chevron-left theme--light"
                  >
                  </i>
                </span>
              </button>
              <div class="v-date-picker-header__value">
                <transition-stub name="tab-transition">
                  <div class="accent--text">
                    <button type="button">
                      March 2021
                    </button>
                  </div>
                </transition-stub>
              </div>
              <button type="button"
                      class="v-btn v-btn--icon v-btn--round theme--light v-size--default"
                      aria-label="Next month"
              >
                <span class="v-btn__content">
                  <i aria-hidden="true"
                     class="v-icon notranslate mdi mdi-chevron-right theme--light"
                  >
                  </i>
                </span>
              </button>
            </div>
            <div class="v-date-picker-table v-date-picker-table--date theme--light">
              <transition-stub name="tab-transition">
                <table>
                  <thead>
                    <tr>
                      <th>
                        S
                      </th>
                      <th>
                        M
                      </th>
                      <th>
                        T
                      </th>
                      <th>
                        W
                      </th>
                      <th>
                        T
                      </th>
                      <th>
                        F
                      </th>
                      <th>
                        S
                      </th>
                    </tr>
                  </thead>
                  <tbody>
                    <tr>
                      <td>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            1
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            2
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            3
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            4
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            5
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            6
                          </div>
                        </button>
                      </td>
                    </tr>
                    <tr>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            7
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            8
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--active v-btn--rounded theme--light accent"
                        >
                          <div class="v-btn__content">
                            9
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            10
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            11
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            12
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            13
                          </div>
                        </button>
                      </td>
                    </tr>
                    <tr>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            14
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            15
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            16
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            17
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            18
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            19
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            20
                          </div>
                        </button>
                      </td>
                    </tr>
                    <tr>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            21
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            22
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            23
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            24
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            25
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            26
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            27
                          </div>
                        </button>
                      </td>
                    </tr>
                    <tr>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            28
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            29
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            30
                          </div>
                        </button>
                      </td>
                      <td>
                        <button type="button"
                                class="v-btn v-btn--text v-btn--rounded theme--light"
                        >
                          <div class="v-btn__content">
                            31
                          </div>
                        </button>
                      </td>
                      <td>
                      </td>
                      <td>
                      </td>
                      <td>
                      </td>
                    </tr>
                  </tbody>
                </table>
              </transition-stub>
            </div>
          </div>
        </transition-stub>
      </div>
      <div class="v-picker__actions v-card__actions">
        <button type="button"
                class="v-btn v-btn--text theme--light v-size--default"
        >
          <span class="v-btn__content">
            Clear
          </span>
        </button>
        <div class="spacer">
        </div>
        <button type="button"
                class="v-btn v-btn--text theme--light v-size--default"
        >
          <span class="v-btn__content">
            Cancel
          </span>
        </button>
        <button type="button"
                class="v-btn v-btn--text theme--light v-size--default primary--text"
        >
          <span class="v-btn__content">
            OK
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
